### ISSUE TITLE
A new lint for imports ending in `::{self};`

### ISSUE THREAD

**Issue by r00ster91** [rust-lang/rust-clippy#6552]
### What it does

It detects imports that end in `::{self};` and emits a warning for them.

### Categories (optional)

- Kind: Perhaps `clippy::style`.

The advantage of the recommended code over the original code is that it's shorter and simpler.

### Drawbacks

None.

### Example

```rust
use std::io::{self};
```

Could be written as:

```rust
use std::io;
```

**Comment by camsteffen** [comment]
I think this is a rustfmt bug. rustfmt does the following:

* `use std::io::{Write};` -> `use std::io::Write;`
* `use std::io::self;` -> `use std::io;`
* `use std::io::{self};` -> `use std::io::{self};` 😲😲😲

**Comment by camsteffen** [comment]
I take it back. See https://github.com/rust-lang/rustfmt/issues/3568. For this lint, clippy would need to check if changing the import will break it.

**Comment by ebobrow** [comment]
This looks interesting! I'd like to give it a try.

@rustbot claim

**Comment by ebobrow** [comment]
I'm just about done with this, but I've run into an issue. As far as I can tell, there's no way to see what items are exported from an external crate, meaning that I can't check if removing `::{self}` from an import would bring in anything else with the same name. I could skip that check for imports from external crates and risk false positives or negatives. Or, am I wrong and there is a way to check?

**Comment by camsteffen** [comment]
For `use a::b::{self}`, I think you should be able to do `tcx.item_children(b_def_id)`.

**Comment by ebobrow** [comment]
Thank you! That works for external crates, but checking a local mod gives:
```
error: internal compiler error: compiler/rustc_middle/src/ty/query/mod.rs:282:1: `tcx.item_children(DefId(0:3 ~ unnecessary_self_imports[317d]::a))` unsupported by its crate; perhaps the `item_children` query was never assigned a provider function

thread 'rustc' panicked at 'Box<Any>', /rustc/07e0e2ec268c140e607e1ac7f49f145612d0f597/library/std/src/panic.rs:59:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
Do you know what that is? If it isn't fixable I can use `tcx.hir().get_if_local()`  for local mods, but I would prefer to have the same check for all imports, if possible.

**Comment by camsteffen** [comment]
Hmm, looks like `item_children` is an "extern" query ([codez](https://github.com/rust-lang/rust/blob/c18c0ad2bc5988ca7953459e5a35ece8e69e35e7/compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs#L34)).

> I can use `tcx.hir().get_if_local()` for local mods

Seems like a good solution.

**Comment by ebobrow** [comment]
Thanks for your help!
